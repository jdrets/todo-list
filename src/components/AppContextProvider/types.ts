import React from 'react'

export interface TasksTypes {
  status: string
  priority: string
  dueDate?: string
  title: string
  description: string
}

export interface AppContextProviderTypes {
  tasks: TasksTypes[]
  setTasks: React.SetStateAction<any>
  newTaskModal: boolean
  setNewTaskModal: React.SetStateAction<any>
}
