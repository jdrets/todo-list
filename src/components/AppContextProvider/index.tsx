import React, { FunctionComponent, useState, createContext, useEffect } from 'react'
import _ from 'lodash'

import TaskService from '../../services/taskService'
import { AppContextProviderTypes } from './types'

export const PageContext = createContext({} as AppContextProviderTypes)

const AppContextProvider: FunctionComponent<any> = ({ children, value = {} }) => {
  const [tasks, setTasks] = useState([])
  const [showCreateTaskModal, setShowCreateTaskModal] = useState(false)

  useEffect(() => {
    const storage = TaskService.get()
    setTasks(storage)
  }, [])

  const updateContextWithNewTask = (task) => {
    const newTasks = _.clone(tasks)
    newTasks.push(task)
    setTasks(_.reverse(newTasks))
  }

  const contextValues = {
    tasks,
    showCreateTaskModal,
    updateContextWithNewTask,
    setShowCreateTaskModal,
    ...value
  }

  return (
    <PageContext.Provider value={contextValues}>{children}</PageContext.Provider>
  )
}

export default AppContextProvider
