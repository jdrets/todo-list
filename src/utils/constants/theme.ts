const palette = {
  purpleDark: '#4B00BC',
  purple: '#6600FF',
  purpleMedium: '#8533FF',
  purpleLight: '#E3D0FF',
  green: '#59BC6F',
  greenLight: '#E6F5E2',
  red: '#EB5757',
  redLight: '#FCE4E4',
  yellow: '#DD9900',
  yellowLight: '#FEF1D8',
  white: '#FFFFFF',
  black: '#000000',
  gray: '#898989',
  grayLight: '#E0E0E0'
}

const theme = {
  colors: {
    primaryColor: palette.purple,
    primaryColorMedium: palette.purpleMedium,
    primaryColorLight: palette.purpleLight,
    primaryColorDark: palette.purpleDark,
    ...palette
  },
  layout: {
    breakpoints: {
      medium: '768px'
    },
    mobile: {
      width: '100%',
      padding: '0 24px'
    },
    desktop: {
      width: '1050px'
    }
  }
}

export default theme
