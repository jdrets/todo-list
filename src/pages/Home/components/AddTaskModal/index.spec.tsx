import React from 'react'
import { ThemeProvider } from 'styled-components'

import { fireEvent, render } from '@testing-library/react'

import { AddTaskModal } from '.'
import theme from '../../../../utils/constants/theme'
import TASK_CONSTANTS from '../../../../utils/constants/task'
import SNACKBAR from '../../../../utils/constants/snackbar'
import AppContextProvider from '../../../../components/AppContextProvider'

const setup = (mockData = {}) => {
  const mock = {
    open: true,
    onClose: jest.fn(),
    showSnackbar: jest.fn(),
    fields: null,
    handleChangeField: jest.fn(),
    fetching: false,
    setFetching: jest.fn(),
    isButtonDisabled: jest.fn(),
    handleCancel: jest.fn()
  }
  return render(
    <ThemeProvider theme={theme}>
      <AppContextProvider>
        <AddTaskModal {...mock} {...mockData} />
      </AppContextProvider>
    </ThemeProvider>
  )
}

describe('<AddTaskModal />', () => {
  test('should match with snapshot', () => {
    const { container } = setup()

    expect(container).toMatchSnapshot()
  })

  test('should fail to create task and call showSnackbar function', () => {
    const showSnackbarMock = jest.fn()
    const mockData = {
      fields: null,
      showSnackbar: showSnackbarMock
    }
    const { getByText } = setup(mockData)
    const submitButton = getByText('Create')

    fireEvent.click(submitButton)

    expect(showSnackbarMock).toHaveBeenCalledWith(SNACKBAR.TYPE.ERROR, SNACKBAR.MESSAGE.ERROR)
  })

  test('should create task correctly and call showSnackbar function', () => {
    const showSnackbarMock = jest.fn()
    const mockData = {
      fields: {
        title: 'dummy title',
        description: 'dummy title',
        status: TASK_CONSTANTS.STATUS.DONE,
        priority: TASK_CONSTANTS.PRIORITY.HIGH
      },
      showSnackbar: showSnackbarMock
    }
    const { getByText } = setup(mockData)
    const submitButton = getByText('Create')

    fireEvent.click(submitButton)

    expect(showSnackbarMock).toHaveBeenCalledWith(SNACKBAR.TYPE.SUCCESS, SNACKBAR.MESSAGE.TASKS.SUCCESS.CREATED)
  })

  test('should text button change when fetching is true', () => {
    const mockData = {
      fetching: true
    }
    const { getByText } = setup(mockData)
    const submitButton = getByText('Creating..')

    expect(submitButton).toBeInTheDocument()
  })
})
