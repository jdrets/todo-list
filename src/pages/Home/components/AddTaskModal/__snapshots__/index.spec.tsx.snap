// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTaskModal /> should match with snapshot 1`] = `
<div>
  <div
    class="sc-hKFyIo kcLwtf"
  >
    <div
      class="sc-eCApGN bRrhpH"
    >
      <div
        class="sc-jSFkmK fZfftN"
      >
        <h2
          class="sc-gKAblj dOnOaT"
        >
          Add new task
        </h2>
        <button
          class="sc-iCoHVE eyyXYf"
          data-testid="modal-close-button"
        >
          <span
            class="sc-dlnjPT feMzXu"
            color="#000000"
            height="22px"
            rotate="0"
            width="22px"
          >
            <svg
              class=""
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              
              <path
                d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="sc-fnVYJo gkpykG"
      >
        <div
          class="sc-fujyUd gEVBzy"
        >
          <div
            class="sc-iqAbSa eEwhLy"
          >
            Status
             
            <span
              class="sc-crzoUp kwOebr"
            >
              *
            </span>
          </div>
          <div
            class="sc-pNWxx eZKhKZ"
          >
            <select
              class="sc-kEqYlL gxXEFH"
              name="status"
              placeholder="Task status"
              required=""
            >
              <option>
                Select task status
              </option>
              <option
                value="To do"
              >
                To do
              </option>
              <option
                value="Doing"
              >
                Doing
              </option>
              <option
                value="Done"
              >
                Done
              </option>
            </select>
            <div
              class="sc-jrsJCI hkbrwn"
            >
              <span
                class="sc-dlnjPT feMzXu"
                color="#000000"
                height="22px"
                rotate="0"
                width="22px"
              >
                <svg
                  class=""
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
                  <path
                    d="M112 184l144 144 144-144"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="48"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-fujyUd gEVBzy"
        >
          <div
            class="sc-iqAbSa eEwhLy"
          >
            Priority
             
            <span
              class="sc-crzoUp kwOebr"
            >
              *
            </span>
          </div>
          <div
            class="sc-pNWxx eZKhKZ"
          >
            <select
              class="sc-kEqYlL gxXEFH"
              name="priority"
              placeholder="Task priority"
              required=""
            >
              <option>
                Select task priority
              </option>
              <option
                value="Low"
              >
                Low
              </option>
              <option
                value="Medium"
              >
                Medium
              </option>
              <option
                value="High"
              >
                High
              </option>
            </select>
            <div
              class="sc-jrsJCI hkbrwn"
            >
              <span
                class="sc-dlnjPT feMzXu"
                color="#000000"
                height="22px"
                rotate="0"
                width="22px"
              >
                <svg
                  class=""
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  
                  <path
                    d="M112 184l144 144 144-144"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="48"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-dIsAE cyGWFl"
        >
          <div
            class="sc-ksluoS fuxgTu"
          >
            Due date
             
          </div>
          <input
            class="sc-bqGHjH lljhnV"
            name="dueDate"
            placeholder="Due date"
            required=""
            type="date"
            value=""
          />
        </div>
        <div
          class="sc-dIsAE cyGWFl"
        >
          <div
            class="sc-ksluoS fuxgTu"
          >
            Title
             
            <span
              class="sc-hBMVcZ cRlzPi"
            >
              *
            </span>
          </div>
          <input
            class="sc-bqGHjH lljhnV"
            name="title"
            placeholder="Task title"
            required=""
            value=""
          />
        </div>
        <div
          class="sc-dIsAE cyGWFl"
        >
          <div
            class="sc-ksluoS fuxgTu"
          >
            Description
             
            <span
              class="sc-hBMVcZ cRlzPi"
            >
              *
            </span>
          </div>
          <textarea
            class="sc-bqGHjH iRBTBZ"
            name="description"
            placeholder="Task description"
            required=""
          />
        </div>
      </div>
      <div
        class="sc-bdnylx fhbphs"
      >
        <div
          class="sc-gtssRu dGLdKs"
        >
          Cancel
        </div>
        <div
          class="sc-gtssRu dSSfSs"
        >
          Create
        </div>
      </div>
    </div>
  </div>
</div>
`;
